### DISCUSSION TITLE
Is it possible to show a message in the index on two lines?

### DISCUSSION THREAD

**Discussion by afh** [neomutt/neomutt#4527] [Development]
In my setup my terminal has about 110 characters with all the information (index, flags, date, sender, list subject, etc.) I'd like to see in neomutt it feels cramped at times.

I'm wondering if it is possible to change the `index_format` so that a message takes up two lines (or more) in the index. For example: the first line showing information about the message index, date, sender, size, etc. and the second line showing the subject.

**Comment by flatcap** [comment]
Unfortunately not (and you're not the first to ask).

Possible **workarounds**, change your index_format / sidebar config:

- Using a `folder-hook`
  If, in certain folders some information is more important  
- Create macros to manually alter your settings
  (depending on your focus at the time)
- Use `%I` (initials) rather than a full name

---

Our `Menu` code that draws the Index is quite basic.
It'd require quite a bit of work to improve it.

A simpler enhancement might be to create a `resize-hook` to automate changing config, e.g.
```
# This doesn't exist, yet
resize-hook 'width < 90' 'set index_format = "ABC"; set sidebar_visible = no'
resize-hook 'width > 89' 'set index_format = "XYZ"; set sidebar_visible = yes'
```

**Comment by afh** [comment]
Thanks for the helpful response with the gerat ideas, @flatcap.

If one were to try and have a look at what would be necessary to support a multi-line per message index view or the mentioned resize-hook enhancement, where would one start best?

**Comment by flatcap** [comment]
The `resize-hook` would be simpler -- we already have plenty of hooks to cut'n'paste and `SIGWINCH` (the signal sent on resize is already caught).

Let me just get a second opinion on the idea, first :-)
@gahr - What do you think of: `resize-hook "CONDITION" "COMMAND"`

Where `"CONDITION"` is `"DIMENSION COMPARISON NUMBER"`, e.g.
- `width > 100`
- `width = 80`
- `height < 40`

**Comment by flatcap** [comment]
As for the multi-line `Menu`s...
It's a great idea, but I hate to think how many 1-line assumptions have been made in the `Menu` code.

I'll keep thinking and have a poke around in the code, later.
I'll see if I can justify how scary it'd be :-)

**Comment by gahr** [comment]
> @gahr - What do you think of: `resize-hook "CONDITION" "COMMAND"`

I like it!

**Comment by flatcap** [comment]
Great.

OK, @afh, I'm still thinking through some of the details, but here's a load to get you started.
This is a **lot** to take in, so have a look through the code, have a think, then ask lots of questions.
You can ask here, or on IRC (mostly during European daylight hours). 

`#neomutt` on `irc.libera.chat` ([web client](https://kiwiirc.com/nextclient/irc.libera.chat/#neomutt))

---

Define the new command, here:
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/mutt_commands.c#L48

And give it a unique id, here:
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/hook.h#L36

Each command has a _parse_ function.
This sample one show some common behaviour:
- `Buffer`s for storage
- `mutt_extract_token()` to split up the string

https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/hook.c#L83

There are some notes on debugging NeoMutt, here: https://neomutt.org/dev/build/debug
Alternatively, you can `mutt_debug()` and run `neomutt -d1`

Once we've split up the string, we'll need to store it somewhere.
NeoMutt keeps a list of Hooks:
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/hook.c#L75

We'll probably create a `Hook` and attach a `Pattern` with our width/height + number details in it.

`struct Hook`
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/hook.c#L64

`struct Pattern`
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/pattern/lib.h#L69

New `Hook`
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/hook.c#L326

It'll take us a while to get there, but at the other end

Signals are intercepted here:
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/mutt_signal.c#L51

and that's noticed in the Index code, here:
https://github.com/neomutt/neomutt/blob/edf5f54b27f07e68d7d7542c0cb6caf0a574f912/index/dlg_index.c#L1262

at that point we'd call a hook function to lookup **our** hook and execute the commands.

**Comment by afh** [comment]
Thanks for all the info @flatcap, I'll make some time to dive in and have a look to see if I fell I could contribute…
